src/mlpack/methods/ann/layer/convolution_impl.hpp

```diff
@@ -302,150 +245,178 @@ template<
     typename ForwardConvolutionRule,
     typename BackwardConvolutionRule,
     typename GradientConvolutionRule,
-    typename InputDataType,
-    typename OutputDataType
+    typename MatType
 >
-void Convolution<
+void ConvolutionType<
     ForwardConvolutionRule,
     BackwardConvolutionRule,
     GradientConvolutionRule,
-    InputDataType,
-    OutputDataType
->::Reset()
+    MatType
+>::SetWeights(typename MatType::elem_type* weightPtr)
 {
-    weight = arma::cube(weights.memptr(), kernelWidth, kernelHeight,
-        outSize * inSize, false, false);
-    bias = arma::mat(weights.memptr() + weight.n_elem,
-        outSize, 1, false, false);
+  MakeAlias(weight, weightPtr, kernelWidth, kernelHeight, maps * inMaps);
+  MakeAlias(bias, weightPtr + weight.n_elem, maps, 1);
+  MakeAlias(weights, weightPtr, weight.n_elem + bias.n_elem, 1);
 }
 
 template<
     typename ForwardConvolutionRule,
     typename BackwardConvolutionRule,
     typename GradientConvolutionRule,
-    typename InputDataType,
-    typename OutputDataType
+    typename MatType
 >
-template<typename eT>
-void Convolution<
+void ConvolutionType<
     ForwardConvolutionRule,
     BackwardConvolutionRule,
     GradientConvolutionRule,
-    InputDataType,
-    OutputDataType
->::Forward(const arma::Mat<eT>& input, arma::Mat<eT>& output)
+    MatType
+>::Forward(const MatType& input, MatType& output)
 {
   batchSize = input.n_cols;
-  arma::cube inputTemp(const_cast<arma::Mat<eT>&>(input).memptr(),
-      inputWidth, inputHeight, inSize * batchSize, false, false);
 
-  if (padWLeft != 0 || padWRight != 0 || padHTop != 0 || padHBottom != 0)
+  // First, perform any padding if necessary.
+  const bool usingPadding =
+      (padWLeft != 0 || padWRight != 0 || padHTop != 0 || padHBottom != 0);
+  const size_t paddedRows = this->inputDimensions[0] + padWLeft + padWRight;
+  const size_t paddedCols = this->inputDimensions[1] + padHTop + padHBottom;
+  if (usingPadding)
   {
-    inputPaddedTemp.set_size(inputTemp.n_rows + padWLeft + padWRight,
-        inputTemp.n_cols + padHTop + padHBottom, inputTemp.n_slices);
-
-    for (size_t i = 0; i < inputTemp.n_slices; ++i)
-    {
-      padding.Forward(inputTemp.slice(i), inputPaddedTemp.slice(i));
-    }
+    // TODO: double-check with a padding test?
```

Comment:
I guess this is still something we should do?

Reply:
Good point---done in 44b30019f.